qhbmlib/inference/qnn.py

```diff
@@ -26,136 +26,167 @@
-  def differentiator(self):
-    return self._differentiator
-
   # TODO(#201): consider Hamiltonian type renaming
-  def _sampled_expectation(self, initial_states: tf.Tensor,
-                           observable: hamiltonian.Hamiltonian):
+  def expectation(self, initial_states: tf.Tensor,
+                  observables: Union[tf.Tensor, hamiltonian.Hamiltonian]):
     """Returns the expectation values of the observables against the QNN.
 
     Args:
       initial_states: Shape [batch_size, num_qubits] of dtype `tf.int8`.
         Each entry is an initial state for the set of qubits.  For each state,
         `qnn` is applied and the pure state expectation value is calculated.
-      observable: Hermitian operator to measure.  Will be tiled to measure
-        the expectation value of the observable in the state
-        `qnn|initial_states[i]>` batched over `i`.
-        Note that since the accepted type is `hamiltonian.Hamiltonian`, the
-        representation is restricted to diagonalized operators.
+      observables: Hermitian operators to measure.  If `tf.Tensor`, strings with
+        shape [n_ops], result of calling `tfq.convert_to_tensor` on a list of
+        cirq.PauliSum, `[op1, op2, ...]`.  Otherwise, a Hamiltonian.  Will be
+        tiled to measure `<op_j>_((qnn)|initial_states[i]>)` for each i and j.
 
     Returns:
-      `tf.Tensor` with shape [batch_size, 1] whose entries are the
-      unaveraged expectation values of `observable` against each transformed
-      initial state.
+      `tf.Tensor` with shape [batch_size, n_ops] whose entries are the
+      unaveraged expectation values of each `operator` against each
+      transformed initial state.
+    """
+    unique_states, idx, _ = utils.unique_bitstrings_with_counts(initial_states)
+    if isinstance(observables, tf.Tensor):
+      total_circuit = self.circuit
+    else:
+      total_circuit = self.circuit + observables.circuit_dagger
+    circuits = total_circuit(unique_states)
+    num_circuits = tf.shape(circuits)[0]
+    tiled_values = tf.tile(
+        tf.expand_dims(total_circuit.symbol_values, 0), [num_circuits, 1])
+    unique_expectations = self._expectation(circuits,
+                                            total_circuit.symbol_names,
+                                            tiled_values, observables)
+    return utils.expand_unique_results(unique_expectations, idx)
+
+  @abc.abstractmethod
+  def _expectation(self, circuits, symbol_names, symbol_values, observables):
+    raise NotImplementedError()
+
+
+class AnalyticQuantumInference(QuantumInference):
+  """Analytic methods for inference on QuantumCircuit objects.
+
+  This class uses the TensorFlow Quantum `Expectation` layer to compute
+  expectation values of observables.  It uses adjoint gradients to compute
+  the derivatives of those expectation values.
+  """
+
+  def __init__(self,
+               input_circuit: circuit.QuantumCircuit,
+               name: Union[None, str] = None):
```

Comment:
could you summarize this piece of context in the docstrings which justifies why we now always use adjoint diff for analytic inference and parameter shift for sampled:

> Specifying alternative differentiators for the 'noiseless' backend, such as ParameterShift, only has the effect of increasing computational overhead, since the expectation values are exact irrespective of the chosen differentiator

Reply:
Done, let me know if it clarifies things